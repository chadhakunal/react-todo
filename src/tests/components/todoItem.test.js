import { shallow, mount } from "enzyme";
import TodoItem from "components/todoItem";
import DoneOutlineIcon from '@mui/icons-material/DoneOutline';
import ClearIcon from '@mui/icons-material/Clear';

const clickFn = jest.fn();

describe('TodoItem', () => {
  const requiredProps = { index: 0, name: "todo", type: "todo", title: "ToDo", description: "World" };

  it('should render correctly with props', () => {
    const component = shallow(<TodoItem {...requiredProps} />);
    expect(component).toMatchSnapshot();
  });

  // Delete
  it('should call onDelete prop when delete clicked', () => {
    const component = mount(<TodoItem onDelete={clickFn} {...requiredProps} />);

    component.find('button#delete-button').simulate('click');

    expect(clickFn).toHaveBeenCalledWith(0, "todo");
    component.unmount();
  });

  // Move
  it('should call moveItem prop when move clicked', () => {
    const component = mount(<TodoItem moveItem={clickFn} {...requiredProps} />);

    component.find('button#move-button').simulate('click');

    expect(clickFn).toHaveBeenCalledWith(0, "todo");
    component.unmount();
  });

  it('should render tick if type is todo', () => {
    const component = shallow(<TodoItem moveItem={clickFn} {...requiredProps} />);
    expect(component.find(DoneOutlineIcon)).toHaveLength(1);
  });

  it('should render cross if type is completed', () => {
    const component = shallow(<TodoItem moveItem={clickFn} {...requiredProps} type="completed" />);
    expect(component.find(ClearIcon)).toHaveLength(1);
  })

  // Edit
  it("Should toggle edit form on edit button click", () => {
    const component = mount(<TodoItem {...requiredProps} />);
    const editForm = component.find('div#edit-form-div');

    expect(editForm.instance().style).toHaveProperty('display', 'none');
    component.find('button#edit-button').simulate('click');
    expect(editForm.instance().style).toHaveProperty('display', 'block');
    component.find('button#edit-button').simulate('click');
    expect(editForm.instance().style).toHaveProperty('display', 'none');

    component.unmount();
  })

  it("Should call onEdit prop on submitting form and toggle the display", () => {
    const component = mount(<TodoItem {...requiredProps} onEdit={clickFn} />);
    component.find('button#edit-form-submit').simulate('submit', { target: [{ value: "title" }, { value: "description" }] });
    expect(clickFn).toHaveBeenCalled();
    component.unmount();
  })
})