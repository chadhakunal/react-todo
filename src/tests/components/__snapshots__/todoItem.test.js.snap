// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should render correctly with props 1`] = `
<div
  className="row align-items-center mt-1"
>
  <div
    className="col-1"
  >
    <Button
      className="icon-button"
      color="secondary"
      id="move-button"
      onClick={[Function]}
      tag="button"
    >
      <Memo(ForwardRef(DoneOutlineIcon))
        fontSize="14px"
      />
    </Button>
  </div>
  <div
    className="col-11"
  >
    <ForwardRef(Accordion)
      className="box"
    >
      <ForwardRef(AccordionSummary)
        aria-controls="panel1a-content"
        expandIcon={
          <Memo
            style={
              Object {
                "color": "rgba(255, 255, 255, 0.5)",
              }
            }
          />
        }
        id="panel1a-header"
      >
        <ForwardRef(Typography)
          className="title"
        >
          ToDo
        </ForwardRef(Typography)>
      </ForwardRef(AccordionSummary)>
      <ForwardRef(AccordionDetails)>
        <div
          className="row"
        >
          <ForwardRef(Typography)
            className="description"
          >
            World
          </ForwardRef(Typography)>
        </div>
        <div
          className="row mt-3"
        >
          <div
            className="col-12"
          >
            <Button
              className="icon-button"
              color="secondary"
              id="delete-button"
              onClick={[Function]}
              tag="button"
            >
              <Memo(ForwardRef(DeleteIcon))
                fontSize="10px"
              />
            </Button>
            <Button
              className="icon-button"
              color="secondary"
              id="edit-button"
              onClick={[Function]}
              tag="button"
            >
              <Memo(ForwardRef(EditIcon))
                fontSize="10px"
              />
            </Button>
          </div>
        </div>
        <div
          className="row p-4"
          id="edit-form-div"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <Form
            id="edit-form"
            onSubmit={[Function]}
            tag="form"
          >
            <FormGroup
              className="mb-2"
              tag="div"
            >
              <Label
                for="title-edit"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Title
              </Label>
              <Input
                id="title-edit"
                name="title"
                type="text"
              />
            </FormGroup>
            <FormGroup
              className="mb-3"
              tag="div"
            >
              <Label
                for="description-edit"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Description
              </Label>
              <Input
                id="description-edit"
                name="description"
                type="textarea"
              />
            </FormGroup>
            <Button
              className="w-100"
              color="secondary"
              id="edit-form-submit"
              tag="button"
              type="submit"
            >
              Save
            </Button>
          </Form>
        </div>
      </ForwardRef(AccordionDetails)>
    </ForwardRef(Accordion)>
  </div>
</div>
`;
